upgrade.py

```diff
@@ -149,41 +149,58 @@ def init_id_maps(src_root, dst_root):
                         logger.error('Failed to parse: ' + line)
                         logger.logException(e)
 
-        # Copy ownership from a path in a source root to another path in a
-        # destination root. The ownership is copied such that it is not
-        # affected by changes in the underlying uid/gid.
-        def copy_ownership(src_root, src_path, dst_root, dst_path):
+        # Copy attributes from a path in a source root to another path in a
+        # destination root. 
+        # - The ownership is copied such that it is notaffected by changes 
+        #   in the underlying uid/gid. Or it's set to the specified.
+        # - The mode is set to the specified.
+        def copy_attributes(src_root, src_path, dst_root, dst_path, attr=None):
+            dst_f = '%s/%s' % (dst_root, dst_path)
+
             st = os.lstat('%s/%s' % (src_root, src_path))
             try:
-                new_uid = dst_uid_map[src_uid_map[st.st_uid]]
-                new_gid = dst_gid_map[src_gid_map[st.st_gid]]
+                user = src_uid_map[st.st_uid]
+                group = src_gid_map[st.st_gid]
+                if attr is not None:
+                    user = attr.get('user', user)
+                    group = attr.get('group', group)
+                new_uid = dst_uid_map[user]
+                new_gid = dst_gid_map[group]
             except IndexError as e:
-                logger.error('Failed to copy ownership')
+                logger.error('Failed to determine ownership')
                 logger.logException(e)
                 return
             if st.st_uid != new_uid or st.st_gid != new_gid:
-                os.lchown('%s/%s' % (dst_root, dst_path), new_uid, new_gid)
+                logger.log("Update uid=%d gid=%d" % (new_uid, new_gid))
```

Comment:
Isn't this a bit too much verbose?

Reply:
I would like to output change detail for future bug check.
Also this logging is very few because most dirs/files won't update ownerships.